### PR TITLE
Support for map of struct

### PR DESCRIPTION
**Pull request by zenthangplus**
I've modified these codes to support for map

- Map of struct
- Map of ptr struct

Related #28

### CODE REVIEW

**Comment by creasty** [comment]
Some test cases are failing.
(CI didn't run on the fork PRs due to insufficient configuration so I run locally.)

```sh-session
$ go test ./...
--- FAIL: TestInit (0.00s)
    --- FAIL: TestInit/map_of_ptr_struct (0.00s)
        defaults_test.go:602: it should set default for Bar field in Struct1 item
        defaults_test.go:605: it should set default for WithDefault field in Struct1 item
        defaults_test.go:608: it should not override Foo field in Struct2 item
        defaults_test.go:614: it should set default for WithDefault field in Struct2 item
FAIL
FAIL    github.com/creasty/defaults     0.095s
?       github.com/creasty/defaults/internal/fixture    [no test files]
FAIL
```

**Comment by zenthangplus** [comment]
@creasty fixed, i've mistake when write new value to map.

File: defaults.go [hunk 1]
```diff
@@ -168,6 +168,31 @@ func setField(field reflect.Value, defaultVal string) error {
...
+				var newVal reflect.Value
+				if v.Kind() == reflect.Ptr {
+					newVal = reflect.ValueOf(ref.Interface())
+				} else {
+					newVal = reflect.ValueOf(ref.Elem().Interface())
+				}
+				field.SetMapIndex(e, newVal)
+			} else {
+				return err
+			}
```
Comment (creasty) [inline defaults.go:194]: Delving in on my side, this feels more elegant and can support even more complex types like `map[string]map[int]Struct`, `map[string]*map[int]Struct`,`map[string]AliasOfSliceOfStruct`, etc.

```go
		for _, e := range field.MapKeys() {
			var v = field.MapIndex(e)

			switch v.Kind() {
			case reflect.Ptr:
				switch v.Elem().Kind() {
				case reflect.Struct, reflect.Slice, reflect.Map:
					if err := setField(v.Elem(), ""); err != nil {
						return err
					}
				}
			case reflect.Struct, reflect.Slice, reflect.Map:
				ref := reflect.New(v.Type())
				ref.Elem().Set(v)
				if err := setField(ref.Elem(), ""); err != nil {
					return err
				}
				field.SetMapIndex(e, ref.Elem().Convert(v.Type()))
			}
		}
```

**Review by creasty — COMMENTED** [review]
Sorry for the delayed response.

**Pull request merged**